free/src/test/scala/cats/free/FreeTTests.scala

```diff
@@ -0,0 +1,218 @@
+        (fu, i) => fu.map(identity)
+      )
+
+    Eq[FreeTListOption[Unit]].eqv(expected, result) should ===(true)
+  }
+
+  test("hoist to universal id equivalent to original instance") {
+    forAll { a: FreeTListOption[Int] =>
+      val b = a.hoist(FunctionK.id)
+      Eq[FreeTListOption[Int]].eqv(a, b) should ===(true)
+    }
+  }
+
+  test("hoist stack-safety") {
+    val a = (0 until 50000).foldLeft(Applicative[FreeTListOption].pure(()))(
+      (fu, i) => fu.flatMap(u => Applicative[FreeTListOption].pure(u))
+    )
+    val b = a.hoist(FunctionK.id)
+  }
+
+  test("transLift for FreeT requires only Functor") {
+    implicit val transLiftInstance = FreeT.catsFreeTransLiftForFreeT[JustFunctor]
+    val d: FreeT[JustFunctor, JustFunctor, Int] = transLiftInstance.liftT[JustFunctor, Int](JustFunctor(1))
+  }
+
+  test("interpret to universal id equivalent to original instance") {
+    forAll { a: FreeTListOption[Int] =>
+      val b = a.interpret(FunctionK.id)
+      Eq[FreeTListOption[Int]].eqv(a, b) should ===(true)
+    }
+  }
+
+  test("interpret stack-safety") {
+    val a = (0 until 50000).foldLeft(Applicative[FreeTListOption].pure(()))(
+      (fu, i) => fu.flatMap(u => Applicative[FreeTListOption].pure(u))
+    )
+    val b = a.interpret(FunctionK.id) // used to overflow
+  }
+
+  test("foldMap consistent with runM") {
+    implicit val listWrapperFunctor = ListWrapper.monad
+    forAll { a: FreeTListOption[Int] =>
+      val x = a.runM(_.list.headOption)
+      val y = a.foldMap(headOption)
+      Eq[Option[Int]].eqv(x, y) should ===(true)
+    }
+  }
+
+  test("== should not return true for unequal instances") {
+    val a = FreeT.pure[List, Option, Int](1).flatMap(x => FreeT.pure(2))
+    val b = FreeT.pure[List, Option, Int](3).flatMap(x => FreeT.pure(4))
+    a == b should be(false)
+  }
+
+  private[free] def liftTUCompilationTests() = {
+    val a: String Xor Int = Xor.right(42)
+    val b: FreeT[Option, String Xor ?, Int] = FreeT.liftTU(a)
+  }
+}
+
+object FreeTTests extends FreeTTestsInstances
+
+sealed trait FreeTTestsInstances {
+
+  import FreeT._
+  import Arbitrary._
+  import org.scalacheck.Arbitrary
+  import cats.kernel.instances.option._
+  import cats.tests.StateTTests._
+  import CartesianTests._
+
+  type IntState[A] = State[Int, A]
+  type FreeTListW[M[_], A] = FreeT[ListWrapper, M, A]
+  type FreeTListWrapper[A] = FreeTListW[ListWrapper, A]
+  type FreeTListOption[A] = FreeTListW[Option, A]
+  type FreeTListState[A] = FreeT[IntState, IntState, A]
+  type MonadRec[F[_]] = Monad[F] with RecursiveTailRecM[F]
+
+  case class JustFunctor[A](a: A)
+
+  implicit val listSemigroupK: SemigroupK[ListWrapper] = ListWrapper.semigroupK
+  
+  implicit val listWrapperMonad: MonadRec[ListWrapper] with Alternative[ListWrapper] = ListWrapper.monadCombine
+
+  implicit val intStateMonad: MonadRec[IntState] = throw("can't summon this instance!")
```

Comment:
@adelbertc @johnynek This is what I was mentioning in gitter. I can't find a way to summon the `StateT` `Monad with RecursiveTailRecM` instance